pkg/datapath/loader/loader.go

```diff
@@ -298,34 +292,44 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 			// this log message should only represent failures with respect to
 			// loading the program.
 			if ctx.Err() == nil {
-				scopedLog.WithError(err).Warn("JoinEP: Failed to attach ingress program")
+				scopedLog.WithError(err).Warnf("JoinEP: Failed to attach (%s) program", progDirection)
 			}
+		}
+		finalizeChan <- finalize
+		errChan <- err
+	}()
+	return finalizeChan, errChan
+}
+
+func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs *directoryInfo) error {
+	// Replace the current program
+	objPath := path.Join(dirs.Output, endpointObj)
+
+	if ep.IsHost() {
+		objPath = path.Join(dirs.Output, hostEndpointObj)
+		if err := l.reloadHostDatapath(ctx, ep, objPath); err != nil {
 			return err
 		}
-		defer finalize()
+	} else {
+		ingressFinalize, ingressErr := asyncReplaceDatapath(ctx, ep, objPath, symbolFromEndpoint, dirIngress)
 
 		if ep.RequireEgressProg() {
-			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, "")
-			if err != nil {
-				scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
-					logfields.Path: objPath,
-					logfields.Veth: ep.InterfaceName(),
-				})
-				// Don't log an error here if the context was canceled or timed out;
-				// this log message should only represent failures with respect to
-				// loading the program.
-				if ctx.Err() == nil {
-					scopedLog.WithError(err).Warn("JoinEP: Failed to attach egress program")
-				}
+			egressFinalize, egressErr := asyncReplaceDatapath(ctx, ep, objPath, symbolToEndpoint, dirEgress)
+			if err := <-egressErr; err != nil {
 				return err
 			}
-			defer finalize()
+			defer (<-egressFinalize)()
 		} else {
 			err := RemoveTCFilters(ep.InterfaceName(), netlink.HANDLE_MIN_EGRESS)
 			if err != nil {
 				log.WithField("device", ep.InterfaceName()).Error(err)
 			}
 		}
+
+		if err := <-ingressErr; err != nil {
```

Comment:
Could https://pkg.go.dev/golang.org/x/sync/errgroup simplify things here? `asyncReplaceDatapath` could take a `chan func() error` it pushes finalizers onto that can be drained in a loop. That would result in a single error check here and tidies things up a bit.

Reply:
Thanks for pointing out, used errgroup here